{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "fb91f56b-fc44-48c1-9345-8916f19d280c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Homework 2 Part 2 (due 7/07/2024)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08469748-4a10-4631-9842-16f74f16fd7f",
   "metadata": {},
   "source": [
    "# Health-care assistance via probabilistic graphical modeling\n",
    "\n",
    "### Objective\n",
    "In this project, you will create a health-care assistance bot that can suggest diagnoses for a set of symptoms based on a probabilistic graphical model.\n",
    "\n",
    "### Step 1: Review \n",
    "Review the code from the Bayesian networks exercise.\n",
    "\n",
    "### Step 2: Acquire data\n",
    "View this [research article](https://www.nature.com/articles/ncomms5212) and download its supplementary data sets 1, 2 and 3. These data sets include the occurrences of diseases, symptoms, and their co-occurrences in the scientific literature. (For the purpose of this exercise, we are going to assume that the frequency of co-occurrences of diseases and symptoms in scientific papers is proportional to the co-occurence frequencies of actual disease cases and symptoms.)\n",
    "\n",
    "### Step 3: Create a Bayesian network\n",
    "Using commands from the `pgmpy` library, create a Bayesian network in which the probability of exhibiting a symptom is conditional on the probability of having an associated disease. \n",
    "\n",
    "### Step 4: Initialize priors\n",
    "Use the disease occurrence data to assign prior probabilities for diseases.\n",
    "\n",
    "### Step 5: Calculate conditional probability tables\n",
    "Use the co-occurrence data to define CPTs for each connected pair of disease and symptoms. (Hint: You may need to assign some occurrences of symptoms to an \"idiopathic disease\" to create valid CPTs.)\n",
    "\n",
    "### Step 6:\n",
    "Create a minimal interface in which your bot asks a users for a list of observed symptoms and then returns the name of the disease that is the most likely match to the symptoms. (Hint: Review the input/output commands that you have used in last week's homework.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ae0991ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pgmpy.models import BayesianNetwork\n",
    "from pgmpy.factors.discrete import TabularCPD\n",
    "from pgmpy.inference import VariableElimination\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "\n",
    "# Load Data\n",
    "sdc = pd.read_csv(\"s_d_coocurrence.txt\", sep='\\t',skiprows=1)\n",
    "sym = pd.read_csv(\"symptom_occurence.txt\", sep='\\t',skiprows=1)\n",
    "dis = pd.read_csv(\"disease_occurence.txt\", sep='\\t',skiprows=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "38e7f11e-ec71-4e73-bbca-b02b632c0850",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44    28042\n",
      "Name: 122226, dtype: int64\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/Documents/GitHub/mathematics-and-ai/mathematics-and-ai/weekly-projects/week2/.venv/lib/python3.12/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:2606\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.Int64HashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:2630\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.Int64HashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 1",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[95], line 87\u001b[0m\n\u001b[1;32m     84\u001b[0m   model\u001b[38;5;241m.\u001b[39maddcpds(CPTs_symptoms)\n\u001b[1;32m     86\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 87\u001b[0m     \u001b[43mdriver\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[95], line 68\u001b[0m, in \u001b[0;36mdriver\u001b[0;34m()\u001b[0m\n\u001b[1;32m     65\u001b[0m little_cpts \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m disease \u001b[38;5;129;01min\u001b[39;00m parents:\n\u001b[1;32m     67\u001b[0m   \u001b[38;5;66;03m# occurrence of the selected disease\u001b[39;00m\n\u001b[0;32m---> 68\u001b[0m   dis_occ \u001b[38;5;241m=\u001b[39m\u001b[43mgrabFromDF\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdis\u001b[49m\u001b[43m,\u001b[49m\u001b[43mdisease\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     70\u001b[0m   sym_occ \u001b[38;5;241m=\u001b[39m grabFromDF(sym,sym\u001b[38;5;241m.\u001b[39miloc[row,\u001b[38;5;241m0\u001b[39m])\n\u001b[1;32m     72\u001b[0m   int_occ \u001b[38;5;241m=\u001b[39m grabFromDF2(sdc,sym\u001b[38;5;241m.\u001b[39miloc[row,\u001b[38;5;241m0\u001b[39m],disease)\n",
      "Cell \u001b[0;32mIn[95], line 7\u001b[0m, in \u001b[0;36mgrabFromDF\u001b[0;34m(df, value, iv_index, dv_index)\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(sub_df):\n\u001b[1;32m      6\u001b[0m   \u001b[38;5;28mprint\u001b[39m(sub_df)\n\u001b[0;32m----> 7\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msub_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m: \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/Documents/GitHub/mathematics-and-ai/mathematics-and-ai/weekly-projects/week2/.venv/lib/python3.12/site-packages/pandas/core/series.py:1121\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1118\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m   1120\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1121\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1123\u001b[0m \u001b[38;5;66;03m# Convert generator to list before going through hashable part\u001b[39;00m\n\u001b[1;32m   1124\u001b[0m \u001b[38;5;66;03m# (We will iterate through the generator there to check for slices)\u001b[39;00m\n\u001b[1;32m   1125\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n",
      "File \u001b[0;32m~/Documents/GitHub/mathematics-and-ai/mathematics-and-ai/weekly-projects/week2/.venv/lib/python3.12/site-packages/pandas/core/series.py:1237\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1234\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1236\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1237\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1239\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/Documents/GitHub/mathematics-and-ai/mathematics-and-ai/weekly-projects/week2/.venv/lib/python3.12/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 1"
     ]
    }
   ],
   "source": [
    "# Code from Dr. Schwarze's announcement\n",
    "def grabFromDF(df,value,iv_index=0,dv_index=1):\n",
    "  \n",
    "  sub_df = (df[df[df.columns[iv_index]]==value][df.columns[dv_index]])\n",
    "  \n",
    "  if len(sub_df):\n",
    "    print(sub_df) # the error originates here I believe, it prints numbers for some reason\n",
    "    return sub_df[1]\n",
    "  else: return 0\n",
    "\n",
    "# Code from Dr. Schwarze's announcement\n",
    "def grabFromDF2(df,value1,value2,iv1_index=0,iv2_index=1,dv_index=2):\n",
    "  var1 = df.columns[iv1_index]\n",
    "  var2 = df.columns[iv2_index]\n",
    "  var3 = df.columns[dv_index]\n",
    "\n",
    "  sub_df = df[df[var1]==value1]\n",
    "  subsub_df = sub_df[sub_df[var2]==value2]\n",
    "\n",
    "  if len(subsub_df):\n",
    "    return subsub_df[var3]\n",
    "  else:\n",
    "    return 0\n",
    "\n",
    "# Code from Dr. Schwarze's announcement\n",
    "def CPT2x2(disease_occ,symptom_occ,interaction_occ,total_disease_occ,total_symptom_occ,total_interaction_occ):\n",
    "  p_disease = disease_occ/total_disease_occ\n",
    "  p_joint = interaction_occ/total_interaction_occ\n",
    "\n",
    "  pTT = (p_joint/p_disease if p_joint>0 else 0.0)\n",
    "  pFT = 1 - pTT\n",
    "  pTF = (symptom_occ-interaction_occ)/total_symptom_occ\n",
    "  pFF = 1 - pTF\n",
    "\n",
    "  return [pFF,pTF,pFT,pTT]\n",
    "\n",
    "def driver():\n",
    "  # Part 3: Define the structure of the Bayesian Network\n",
    "  model = BayesianNetwork()\n",
    "  # add disease nodes with occurrence over 500\n",
    "  for row in range(dis.shape[0]):\n",
    "    if dis.iloc[row,1] > 500:\n",
    "      model.add_node(dis.iloc[row,0])\n",
    "\n",
    "  # add symptom nodes and edges (with some failsafes)\n",
    "  for row in range(sdc.shape[0]):\n",
    "    if sdc.iloc[row,1] in model:\n",
    "      model.add_node(sdc.iloc[row,0])\n",
    "      if (sdc.iloc[row,1] != sdc.iloc[row,0]) and not(nx.has_path(model,sdc.iloc[row,0],sdc.iloc[row,1])):\n",
    "        model.add_edge(sdc.iloc[row,1],sdc.iloc[row,0])\n",
    "\n",
    "  # Part 4: Initialize Prior Probabilities (assuming user has a disease in the list)\n",
    "  tot_sym_occ = sum(sym.iloc[:,1])\n",
    "  tot_dis_occ = sum(dis.iloc[:,1]) # find total occurrences\n",
    "  tot_int_occ = sum(sdc.iloc[:,2])\n",
    "\n",
    "  # calculate prior probabilities\n",
    "  for row in range(dis.shape[0]):\n",
    "    if dis.iloc[row,0] in model:\n",
    "      p = dis.iloc[row,1]/tot_dis_occ # probability has the disease\n",
    "      cpd_d = TabularCPD(variable=dis.iloc[row,0],variable_card=2,values=[[1-p],[p]])\n",
    "      model.add_cpds(cpd_d)\n",
    "\n",
    "  # Part 5: Define CPTs for symptom nodes\n",
    "  CPTs_symptoms = []\n",
    "  for row in range(sym.shape[0]):\n",
    "  # get all parent nodes\n",
    "    if sym.iloc[row,0] in model:\n",
    "      parents = list(model.predecessors(sym.iloc[row,0]))\n",
    "\n",
    "      little_cpts = []\n",
    "      for disease in parents:\n",
    "        # occurrence of the selected disease\n",
    "        dis_occ =grabFromDF(dis,disease)\n",
    "        \n",
    "        sym_occ = grabFromDF(sym,sym.iloc[row,0])\n",
    "        \n",
    "        int_occ = grabFromDF2(sdc,sym.iloc[row,0],disease)\n",
    "        \n",
    "        little_cpt = CPT2x2(dis_occ,sym_occ,int_occ,tot_dis_occ,tot_sym_occ,tot_int_occ)\n",
    "        little_cpts += [little_cpt]\n",
    "      rowT = []\n",
    "      for bool_combo in itertools.product([0,1], repeat=len(parents)):\n",
    "        cond_probs = [little_cpts[i][2+b] for i, b in enumerate(bool_combo)]\n",
    "        rowT += [np.prod(cond_probs)]\n",
    "      rowF = [1-val for val in rowT]\n",
    "\n",
    "      cpt = TabularCPD(variable=sym.iloc[row,0],variable_card=2,values=[rowF,rowT],evidence=parents,evidence_card=[2 for _ in parents])\n",
    "      CPTs_symptoms += [cpt]\n",
    "  model.addcpds(CPTs_symptoms)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    driver()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45f444f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an inference object\n",
    "inference = VariableElimination(model)\n",
    "\n",
    "# Compute the probability of D given A=1 and B=0\n",
    "prob_d_given_a1_b0 = inference.query(variables=['D'], evidence={'A': 1, 'B': 0})\n",
    "print(prob_d_given_a1_b0)\n",
    "\n",
    "# Compute the marginal probability of C\n",
    "prob_c = inference.query(variables=['C'])\n",
    "print(prob_c)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
